### PR TITLE
Feature/fallback

### PR DESCRIPTION
**Pull request by zeroknots**
Adding prototype for function sig specificed fallbacks

### CODE REVIEW

File: src/core/ModuleManager.sol [hunk 1]
```diff
@@ -144,17 +148,20 @@ abstract contract ModuleManager is AccountBase, Receiver {
...
-    function _installFallbackHandler(address handler, bytes calldata initData) internal virtual {
-        if (_isFallbackHandlerInstalled()) revert();
-        _getModuleManagerStorage().fallbackHandler = handler;
+    function _installFallbackHandler(address handler, bytes calldata params) internal virtual {
+        (bytes4 functionSig, CallType calltype, bytes memory initData) =
+            abi.decode(params, (bytes4, CallType, bytes));
+        if (_isFallbackHandlerInstalled(functionSig)) revert();
+
+        $moduleManager().$fallbacks[functionSig] = FallbackHandler(handler, calltype);
         IFallback(handler).onInstall(initData);
```
Comment (leekt) [inline src/core/ModuleManager.sol:165]: if calltype is delegatecall, should `onInstall()` be called with delegatecall?

Comment (zeroknots) [reply src/core/ModuleManager.sol:165]: good shout!

File: src/core/ModuleManager.sol [hunk 2]
```diff
@@ -165,58 +172,105 @@ abstract contract ModuleManager is AccountBase, Receiver {
...
-            // memory is used, therefore we need to guarantee memory safety (keeping the free memory
-            // point 0x40 slot intact,
-            // not going beyond the scratch space, etc)
-            // Solidity docs: https://docs.soliditylang.org/en/latest/assembly.html#memory-safety
-            function allocate(length) -> pos {
-                pos := mload(0x40)
-                mstore(0x40, add(pos, length))
+        FallbackHandler storage $fallbackHandler = $moduleManager().$fallbacks[msg.sig];
+        address handler = $fallbackHandler.handler;
+        CallType calltype = $fallbackHandler.calltype;
```
Comment (nlordell) [inline src/core/ModuleManager.sol:212]: Just out of curiosity, what does the Solidity compiler do here? Does it know to optimize this into a single SLOAD?

Comment (zeroknots) [reply src/core/ModuleManager.sol:212]: If I remember correctly it does. But worth double checking. or writing in `assembly`.

Will do and let you know :)

Comment (kopy-kat) [reply src/core/ModuleManager.sol:212]: didnt investigate properly, but the gas diff when hardcoding the handler is `92 gas` so it seems like there is just one SLOAD

**Pull request merged**